server/transaction/manager.go

```diff
@@ -59,20 +59,44 @@ type Tx interface {
 	Rollback(ctx context.Context) error
 }
 
+type stagedDatabase interface {
+	IsMetadataChange() bool
+}
+
 // SessionCtx is used to store any baggage for the lifetime of the transaction. We use it to stage the database inside
 // a transaction when the transaction is performing any DDLs.
 type SessionCtx struct {
-	db any
+	db                    stagedDatabase
+	noMetadataChanged     *bool
+	metadataChangeSession bool
 }
 
-func (c *SessionCtx) StageDatabase(db any) {
+func (c *SessionCtx) MetadataChangeSession(mcs bool) {
+	c.metadataChangeSession = mcs
+}
+
+func (c *SessionCtx) StageDatabase(db stagedDatabase) {
 	c.db = db
 }
 
-func (c *SessionCtx) GetStagedDatabase() any {
+func (c *SessionCtx) GetStagedDatabase() stagedDatabase {
 	return c.db
 }
 
+func (c *SessionCtx) MarkNoMetadataStateChanged() {
+	ptrFalse := false
+	c.noMetadataChanged = &ptrFalse
+}
+
+func (c *SessionCtx) IsMetadataStateChanged() bool {
+	incrMetadataVersion := (c.db == nil && c.metadataChangeSession) || (c.db != nil && c.db.IsMetadataChange())
+	if c.noMetadataChanged != nil {
+		incrMetadataVersion = *c.noMetadataChanged
```

Comment:
noMetadataChanged == true will return incrMetadataVersion == true?

Reply:
it should never be true but let me add a guard against it.